packages/block-editor/src/components/writing-flow/use-select-all.js

```diff
@@ -7,38 +7,62 @@ import { first, last } from 'lodash';
  * WordPress dependencies
  */
 import { isEntirelySelected } from '@wordpress/dom';
-import { isKeyboardEvent } from '@wordpress/keycodes';
+import { useRef, useCallback } from '@wordpress/element';
 import { useSelect, useDispatch } from '@wordpress/data';
-import { useRefEffect } from '@wordpress/compose';
+import { useShortcut } from '@wordpress/keyboard-shortcuts';
 
 /**
  * Internal dependencies
  */
 import { store as blockEditorStore } from '../../store';
 
 export default function useSelectAll() {
-	const { getBlockOrder } = useSelect( blockEditorStore );
+	const ref = useRef();
+	const {
+		getBlockOrder,
+		getSelectedBlockClientIds,
+		getBlockRootClientId,
+	} = useSelect( blockEditorStore );
 	const { multiSelect } = useDispatch( blockEditorStore );
 
-	return useRefEffect( ( node ) => {
-		function onKeyDown( event ) {
-			if (
-				isKeyboardEvent.primary( event, 'a' ) &&
-				isEntirelySelected( event.target )
-			) {
-				const blocks = getBlockOrder();
-				const firstClientId = first( blocks );
-				const lastClientId = last( blocks );
-				if ( firstClientId !== lastClientId ) {
-					multiSelect( firstClientId, lastClientId );
-					event.preventDefault();
-				}
-			}
+	const callback = useCallback( ( event ) => {
+		const selectedClientIds = getSelectedBlockClientIds();
+
+		if ( ! selectedClientIds.length ) {
+			return;
+		}
+
+		if (
+			selectedClientIds.length === 1 &&
+			! isEntirelySelected( event.target )
+		) {
+			return;
+		}
+
+		const [ firstSelectedClientId ] = selectedClientIds;
+		const rootClientId = getBlockRootClientId( firstSelectedClientId );
+		let blockClientIds = getBlockOrder( rootClientId );
+
+		if ( selectedClientIds.length === blockClientIds.length ) {
+			blockClientIds = getBlockOrder(
+				getBlockRootClientId( rootClientId )
+			);
```

Comment:
A comment linking to this PR would be a nice bonus to explain this little sleight of hand:

```suggestion
		// If we have selected all sibling nested blocks, try selecting up a level.
		// See: https://github.com/WordPress/gutenberg/pull/31859/
		if ( selectedClientIds.length === blockClientIds.length ) {
			blockClientIds = getBlockOrder(
				getBlockRootClientId( rootClientId )
			);
```